### PR TITLE
Sort block improvements

### PR DESCRIPTION
**Pull request by kitaisreal**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Merge PR https://github.com/ClickHouse/ClickHouse/pull/46558 (Avoid processing already sorted data). Avoid block permutation during sort if the block is already sorted.

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
This is an automated comment for commit 9d5840c1064dbb32c5e6f210c2166fdc885f8bc4 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/50697/9d5840c1064dbb32c5e6f210c2166fdc885f8bc4/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/Interpreters/sortBlock.cpp [hunk 1]
```diff
@@ -240,33 +310,28 @@ bool isAlreadySorted(const Block & block, const SortDescription & description)
     if (!block)
         return true;
 
-    size_t rows = block.rows();
-
     ColumnsWithSortDescriptions columns_with_sort_desc = getColumnsWithSortDescription(block, description);
+    bool is_collation_required = false;
```
Comment (hanfei1991) [inline src/Interpreters/sortBlock.cpp:314]: Why did we not consider collation before? Is it a bug?

**Comment by hanfei1991** [comment]
`02497_storage_file_reader_selection` is likely a flaky test

**Review by hanfei1991 — APPROVED** [review]
All the failed tests are unrelate, (although there're many failure ...)

File: src/Interpreters/sortBlock.cpp (hunk 1 above)
Comment (kitaisreal) [reply src/Interpreters/sortBlock.cpp:314]: It is not a bug. It was mentioned in documentation that collation is not supported for this method.
I added collation support and updated documentation.

**Pull request merged**